### DISCUSSION TITLE
trouble running docker GemNotFound Gemfile.lock

### DISCUSSION THREAD

**Discussion by korinagoldsmith** [alshedivat/al-folio#2724] [Q&A]
I've been having trouble running docker for a while now. This is the error I keep getting. I tried manually updating to the current repo using git rebase but I'm still having errors.

2024-09-26 14:31:50 jekyll-1  | error: pathspec 'Gemfile.lock' did not match any file(s) known to git
2024-09-26 20:30:45 jekyll-1  | Resolving dependencies...
2024-09-26 20:30:45 jekyll-1  | /usr/local/bundle/gems/bundler-2.5.20/lib/bundler/definition.rb:599:in `materialize': Could not find jekyll-diagrams-0.10.0, mini_racer-0.6.4, faraday-retry-2.2.1, libv8-node-16.19.0.1-aarch64-linux, faraday-2.12.0, faraday-net_http-3.3.0 in locally installed gems (Bundler::GemNotFound)

I also have a build error. I will insert it below if that helps.

24.58 Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
24.58 
24.58 current directory:
24.58 /var/lib/gems/3.2.0/gems/mini_racer-0.16.0/ext/mini_racer_extension
24.58 /usr/bin/ruby3.2 -I/usr/lib/ruby/vendor_ruby extconf.rb
24.58 checking for -lpthread... yes
24.58 checking for whether -Wl,--exclude-libs=ALL  is accepted as LDFLAGS... no
24.58 creating Makefile
24.58 
24.58 current directory:
24.58 /var/lib/gems/3.2.0/gems/mini_racer-0.16.0/ext/mini_racer_extension
24.58 make DESTDIR\= sitearchdir\=./.gem.20240926-7-h5zbiw
24.58 sitelibdir\=./.gem.20240926-7-h5zbiw clean
24.58 
24.58 current directory:
24.58 /var/lib/gems/3.2.0/gems/mini_racer-0.16.0/ext/mini_racer_extension
24.58 make DESTDIR\= sitearchdir\=./.gem.20240926-7-h5zbiw
24.58 sitelibdir\=./.gem.20240926-7-h5zbiw
24.58 compiling mini_racer_extension.cc
24.58 linking shared-object mini_racer_extension.so
24.58 /usr/bin/ld: cannot find
24.58 /var/lib/gems/3.2.0/gems/libv8-node-18.19.0.0-aarch64-linux/vendor/v8/aarch64-linux-gnu/libv8/obj/libv8_monolith.a:
24.58 No such file or directory
24.58 collect2: error: ld returned 1 exit status
24.58 make: *** [Makefile:265: mini_racer_extension.so] Error 1
24.58 
24.58 make failed, exit code 2
24.58 
24.58 Gem files will remain installed in /var/lib/gems/3.2.0/gems/mini_racer-0.16.0
24.58 for inspection.
24.58 Results logged to
24.58 /var/lib/gems/3.2.0/extensions/aarch64-linux-gnu/3.2.0/mini_racer-0.16.0/gem_make.out
24.58 
24.58   /usr/lib/ruby/vendor_ruby/rubygems/ext/builder.rb:119:in `run'
24.58   /usr/lib/ruby/vendor_ruby/rubygems/ext/builder.rb:53:in `block in make'
24.58   /usr/lib/ruby/vendor_ruby/rubygems/ext/builder.rb:45:in `each'
24.58   /usr/lib/ruby/vendor_ruby/rubygems/ext/builder.rb:45:in `make'
24.58   /usr/lib/ruby/vendor_ruby/rubygems/ext/ext_conf_builder.rb:42:in `build'
24.58   /usr/lib/ruby/vendor_ruby/rubygems/ext/builder.rb:187:in `build_extension'
24.58 /usr/lib/ruby/vendor_ruby/rubygems/ext/builder.rb:221:in `block in
24.58 build_extensions'
24.58   /usr/lib/ruby/vendor_ruby/rubygems/ext/builder.rb:218:in `each'
24.58   /usr/lib/ruby/vendor_ruby/rubygems/ext/builder.rb:218:in `build_extensions'
24.58   /usr/lib/ruby/vendor_ruby/rubygems/installer.rb:844:in `build_extensions'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/rubygems_gem_installer.rb:109:in
24.58 `build_extensions'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/rubygems_gem_installer.rb:28:in
24.58 `install'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/source/rubygems.rb:205:in
24.58 `install'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/installer/gem_installer.rb:55:in
24.58 `install'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/installer/gem_installer.rb:17:in
24.58 `install_from_spec'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/installer/parallel_installer.rb:133:in
24.58 `do_install'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/installer/parallel_installer.rb:124:in
24.58 `block in worker_pool'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/worker.rb:62:in
24.58 `apply_func'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/worker.rb:57:in `block in
24.58 process_queue'
24.58   /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/worker.rb:54:in `loop'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/worker.rb:54:in
24.58 `process_queue'
24.58 /var/lib/gems/3.2.0/gems/bundler-2.5.20/lib/bundler/worker.rb:90:in `block (2
24.58 levels) in create_threads'
24.58 
24.58 An error occurred while installing mini_racer (0.16.0), and Bundler cannot
24.58 continue.
24.58 
24.58 In Gemfile:
24.58   mini_racer
------
Dockerfile:35
--------------------
  33 |     WORKDIR /srv/jekyll
  34 |     
  35 | >>> RUN bundle install --no-cache
  36 |     # && rm -rf /var/lib/gems/3.1.0/cache
  37 |     EXPOSE 8080
--------------------
ERROR: failed to solve: process "/bin/sh -c bundle install --no-cache" did not complete successfully: exit code: 5

**Comment by riasat-sheikh** [comment]
I am also having some trouble running `docker compose up`. I get the following error which is suggesting that the docker image doesn't include those gems or something? Maybe I understood wrong!!


For now, I am back to [the legacy method!!](https://github.com/alshedivat/al-folio/blob/master/INSTALL.md#local-setup-legacy-no-longer-supported)
```shell
[+] Running 1/0
 ✔ Container thefermi0ngithubio-jekyll-1  Created                                                                                                                                                                 0.0s 
Attaching to jekyll-1
jekyll-1  | /usr/local/bundle/gems/bundler-2.5.7/lib/bundler/definition.rb:596:in `materialize': Could not find jekyll-4.3.3, jekyll-toc-0.18.0, mini_racer-0.12.0, webrick-1.8.1, css_parser-1.17.1, addressable-2.8.6, i18n-1.14.5, rouge-4.2.1, uglifier-4.2.0, nokogiri-1.16.5-aarch64-linux, libv8-node-21.7.2.0-aarch64-linux, public_suffix-5.0.5, concurrent-ruby-1.2.3, sass-embedded-1.77.0-aarch64-linux-gnu, rexml-3.2.8, unicode-display_width-2.5.0, activesupport-7.1.3.2, google-protobuf-4.26.1-aarch64-linux, rb-inotify-0.10.1, strscan-3.1.0, minitest-5.22.3, ffi-1.16.3 in locally installed gems (Bundler::GemNotFound)
jekyll-1  |     from /usr/local/bundle/gems/bundler-2.5.7/lib/bundler/definition.rb:203:in `specs'
jekyll-1  |     from /usr/local/bundle/gems/bundler-2.5.7/lib/bundler/definition.rb:270:in `specs_for'
jekyll-1  |     from /usr/local/bundle/gems/bundler-2.5.7/lib/bundler/runtime.rb:18:in `setup'
jekyll-1  |     from /usr/local/bundle/gems/bundler-2.5.7/lib/bundler.rb:162:in `setup'
jekyll-1  |     from /usr/local/bundle/gems/jekyll-4.3.4/lib/jekyll/plugin_manager.rb:52:in `require_from_bundler'
jekyll-1  |     from /usr/local/bundle/gems/jekyll-4.3.4/exe/jekyll:11:in `<top (required)>'
jekyll-1  |     from /usr/local/bundle/bin/jekyll:25:in `load'
jekyll-1  |     from /usr/local/bundle/bin/jekyll:25:in `<main>'
```


Details
- macOS: Sequoia 15.0 (24A335)
- docker desktop: 4.34.2 (167172)
- docker compose: version v2.29.2-desktop.2

**Reply by george-gca** [reply]
@pourmand1376 maybe this have something to do with the `entry_point.sh` update?

**Comment by pourmand1376** [comment]
Hey Guys. 

I've fixed the bug so that you won't ever see `Gemfile.lock` git problem. But the problem doesn't solve there. 

I found and tested both of your repositories at [here](https://github.com/korinagoldsmith/freshman-portfolio). I found out that your docker versions are very different from our built ones (some gem packages are deleted from repo). @korinagoldsmith 

For this to solve, we may have to take three solutions:
- You can solve this easily by yourself by using `docker compose build` and `docker compose up`. It would built a new docker image for you. You may go ahead from there.
- You may update your repository to the upstream repository so that you would have the same dependencies. 
- We should version whenever we change dependencies, so the users would somehow have access to previous versions. I have to think about this more. 

To solve @TheFermi0n, I think `docker compose build` and `docker compose up` would solve the problem also.

**Reply by pourmand1376** [reply]
Specifically, I think changing `Gemfile` should be always be with a release. So that we would have both previous Docker Image and New docker image in the Repo. 

We just have to provide the access for the users. Maybe changing the version from `latest` to real-version may solve this issue in the future as we are currently not using versioning the good way. We are just archiving stuff. 

https://github.com/alshedivat/al-folio/blob/c20074c8cab8df2050350e2367482fdd5328a08e/docker-compose.yml#L2-L5

**Reply by riasat-sheikh** [reply]
> To solve @TheFermi0n, I think `docker compose build` and `docker compose up` would solve the problem also.

Thank you very much for such a quick fix. Running `pull`, `build` and `up` in sequence worked properly.

**Reply by george-gca** [reply]
> Specifically, I think changing `Gemfile` should be always be with a release. So that we would have both previous Docker Image and New docker image in the Repo. 

I agree. 


> We just have to provide the access for the users. Maybe changing the version from `latest` to real-version may solve this issue in the future as we are currently not using versioning the good way. We are just archiving stuff. 

I agree. Also I think we shouldn't have a release schedule, rather make releases as they happen naturally. For instance, once I [had to update the dependencies](https://github.com/alshedivat/al-folio/pull/2709) because of a security vulnerability in one of them. This should trigger a x.x.+1 version.

**Reply by pourmand1376** [reply]
Definitely agree. Releases should be natural not time-based.

**Reply by george-gca** [reply]
@alshedivat what do you think?

**Reply by jaiarora0011** [reply]
For me, doing `docker compose build` and `docker compose up` solves the problem locally, but I get this error when pushing the `Gemfile.lock` changes to github:
```
Downloading rexml-3.3.6 revealed dependencies not in the API or the lockfile
(strscan (>= 0)).
Running `bundle update rexml` should fix the problem.
Error: The process '/opt/hostedtoolcache/Ruby/3.2.2/x64/bin/bundle' failed with exit code 34
```

**Reply by pourmand1376** [reply]
You just shouldn't push Gemfile changes to remote. 

We have done much in the new version so that people can't push changes of this file to github.

**Reply by riasat-sheikh** [reply]
> We have done much in the new version so that people can't push changes of this file to github.

Thank you very much. That would be great. Right now I need to manually revert or deselect the gemlock changes(file).
